packages/@uppy/companion-client/src/RequestClient.js

```diff
@@ -94,92 +89,80 @@ export default class RequestClient {
-    return (err) => {
-      if (!err?.isAuthError) {
-        // eslint-disable-next-line no-param-reassign
-        err = new ErrorWithCause(`Could not ${method} ${this.#getUrl(path)}`, { cause: err })
+  /*
+    Preflight was added to avoid breaking change between older Companion-client versions and
+    newer Companion versions and vice-versa. Usually the break will manifest via CORS errors because a
+    version of companion-client could be sending certain headers to a version of Companion server that
+    does not support those headers. In which case, the default preflight would lead to CORS.
+    So to avoid those errors, we do preflight ourselves, to see what headers the Companion server
+    we are communicating with allows. And based on that, companion-client knows what headers to
+    send and what headers to not send.
+
+    The preflight only happens once throughout the life-cycle of a certain
+    Companion-client <-> Companion-server pair (allowedHeadersCache).
+    Subsequent requests use the cached result of the preflight.
+  */
+  async preflight (path) {
+    const allowedHeadersCached = allowedHeadersCache.get(this.hostname)
+    if (allowedHeadersCached != null) return allowedHeadersCached
+
+    const promise = (async () => {
+      try {
+        const response = await fetch(this.#getUrl(path), { method: 'OPTIONS' })
+
+        const header = response.headers.get('access-control-allow-headers')
+        if (header == null) return undefined
+
+        this.uppy.log(`[CompanionClient] adding allowed preflight headers to companion cache: ${this.hostname} ${header}`)
+
+        const allowedHeaders = header.split(',').map((headerName) => headerName.trim().toLowerCase())
+        allowedHeadersCache.set(this.hostname, allowedHeaders)
+        return allowedHeaders
+      } catch (err) {
+        this.uppy.log(`[CompanionClient] unable to make preflight request ${err}`, 'warning')
+        return undefined
       }
-      return Promise.reject(err)
-    }
-  }
+    })()
 
-  preflight (path) {
-    if (this.preflightDone) {
-      return Promise.resolve(this.allowedHeaders.slice())
-    }
+    allowedHeadersCache.set(this.hostname, promise)
 
-    return fetch(this.#getUrl(path), {
-      method: 'OPTIONS',
-    })
-      .then((response) => {
-        if (response.headers.has('access-control-allow-headers')) {
-          this.allowedHeaders = response.headers.get('access-control-allow-headers')
-            .split(',').map((headerName) => headerName.trim().toLowerCase())
-        }
-        this.preflightDone = true
-        return this.allowedHeaders.slice()
-      })
-      .catch((err) => {
-        this.uppy.log(`[CompanionClient] unable to make preflight request ${err}`, 'warning')
-        this.preflightDone = true
-        return this.allowedHeaders.slice()
-      })
+    const allowedHeadersNew = await promise
+    const fallbackAllowedHeaders = ['accept', 'content-type', 'uppy-auth-token']
```

Comment:
good catch! maybe set it here instead? then only need one line instead of three lines?
```suggestion
    allowedHeadersCache.set(this.hostname, allowedHeadersNew)
    const fallbackAllowedHeaders = ['accept', 'content-type', 'uppy-auth-token']
```

Reply:
You mean this?
```suggestion
    const fallbackAllowedHeaders = ['accept', 'content-type', 'uppy-auth-token']
    allowedHeadersCache.set(this.hostname, allowedHeadersNew ?? fallbackAllowedHeaders)
```

Yes that would work, we could also define `const allowedHeadersNew = (await promise) ?? fallbackAllowedHeaders` to avoid to use twice the null coalescing operator.

Reply:
I'm not sure what the is right thing to do. I see that the current code will set default headers OPTIONS request fails, however I feel like that might cause incorrect behaviour, if the user gets a network error or similar, and we should try again to fetch OPTIONS? hence setting the cache to undefined if it fails so it will try again next time.

Reply:
Sure sournds good, you need to update the code and the comment to reflect that though.